### ISSUE TITLE
[preformence ]support low precision sampling

### ISSUE THREAD

**Issue by yonikremer** [flashinfer-ai/flashinfer#1903]
Hey, I am using a 16 bit floating point model that outputs a 16 bit floating point logits.
I noticed that when in `flashinfer.sampling.top_p_sampling_from_probs` the logits are converted into a higher presicion tensor, which interduces large memory overhead without imporing precision (the original data allready have low presion).
Moreover, the internal cuda kernls are generic over the data type.

I will be wiiling to contribute.

**Comment by yzh119** [comment]
Hi @yonikremer , yes we can modify the kernel to support bf16/fp16 input (that's actually our original design years ago but unfortunately we found them highly numerical unstable).

Anyways it's something we can add back. Let me know if you need any guidance how to implement this!

**Comment by yonikremer** [comment]
How can I check the numeric stability of the kernel?

**Comment by yonikremer** [comment]
I am also adding double (64bit) support. What do you think about 8 bit precision? I suspect it might be to low for stable sampling.
